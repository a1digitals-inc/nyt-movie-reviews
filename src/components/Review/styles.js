import styled from 'styled-components';
import { darken } from 'polished';

export const Container = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding: 20px;
  margin-top: 15px;
  border-radius: 4px;
  background: #fff;

  &:hover {
    background: ${darken(0.04, '#fff')};
  }

  div {
    strong {
      display: block;
      font-size: 20px;
    }

    small {
      display: flex;
      align-items: stretch;
      padding: 0;
      margin: 2px 0 0 0;
      color: #999;

      img {
        width: 16px;
        height: 16px;
        padding: 0;
        margin: 0 4px 0 0;
      }
    }

    span {
      display: block;
      margin-top: 15px;
      color: #000000;
    }

    a {
      display: block;
      margin-top: 15px;
      color: #999;
      font-weight: bold;

      &:hover {
        color: ${darken(0.5, '#999')};
      }
    }

    button {
      display: block;
      margin-top: 5px;
      color: #999;
      background-color: transparent;
      font-weight: bold;
      border: none;
      cursor: pointer;

      &:hover {
        color: ${darken(0.5, '#999')};
      }
    }
  }

  img {
    margin: 0 0 0 20px;
    height: 120px;
  }
`;
