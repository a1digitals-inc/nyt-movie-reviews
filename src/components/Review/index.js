import React, { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import PropTypes from 'prop-types';
import { format } from 'date-fns';

import {
  addToFavorites,
  removeFromFavorites,
} from '../../store/modules/favorites/actions';
import { Container } from './styles';
import checkLogo from '../../assets/images/check.svg';
import emptyImageLogo from '../../assets/images/empty-image.svg';

const Review = props => {
  const { rawData } = props;

  const {
    display_title: title,
    summary_short: description,
    publication_date: date,
    multimedia,
    link: { url },
    byline: criticName,
    critics_pick: criticsPick,
  } = rawData;

  const dispatch = useDispatch();

  const [data] = useState({
    title,
    description,
    date,
    image: multimedia && multimedia.src ? multimedia.src : emptyImageLogo,
    url,
    criticName,
    criticsPick: !!criticsPick,
  });

  function handleAddFavorites() {
    dispatch(addToFavorites(rawData));
  }

  function handleRemoveFavorites() {
    dispatch(removeFromFavorites(data.title, data.criticName));
  }

  const isFavorite = useSelector(state => {
    return state.favorites.find(favorite => {
      return (
        favorite.display_title === data.title &&
        favorite.byline === data.criticName
      );
    });
  });

  return (
    <Container>
      <div>
        <strong>{data.title}</strong>
        {data.criticsPick && (
          <small>
            <img src={checkLogo} alt="NYTCriticsPick" />
            NYT Critic Pick
          </small>
        )}
        <small>{format(new Date(data.date), 'P')}</small>
        <small>By {data.criticName}</small>
        <span>{data.description}</span>
        <a href={data.url} target="_blank" rel="noopener noreferrer">
          Read review
        </a>
        {isFavorite ? (
          <button
            className="handleFavoritesButton"
            type="button"
            onClick={handleRemoveFavorites}
          >
            Remove from favorites
          </button>
        ) : (
          <button
            className="handleFavoritesButton"
            type="button"
            onClick={handleAddFavorites}
          >
            Add to favorites
          </button>
        )}
      </div>
      <img src={data.image} alt="MovieImage" />
    </Container>
  );
};

Review.propTypes = {
  rawData: PropTypes.shape({
    display_title: PropTypes.string.isRequired,
    summary_short: PropTypes.string.isRequired,
    publication_date: PropTypes.string.isRequired,
    multimedia: PropTypes.shape({
      src: PropTypes.string,
    }),
    link: PropTypes.shape({
      url: PropTypes.string.isRequired,
    }),
    byline: PropTypes.string.isRequired,
    critics_pick: PropTypes.number,
  }).isRequired,
};

export default Review;
