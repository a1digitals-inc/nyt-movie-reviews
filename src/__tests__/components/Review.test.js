import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';

import Review from '../../components/Review';

describe('Review component', () => {
  const middlewares = [];
  const mockStore = configureStore(middlewares);
  const initialState = {
    favorites: [],
  };
  const store = mockStore(initialState);

  it('should render without crashing', () => {
    const props = {
      rawData: {
        display_title: 'title',
        summary_short: 'description',
        publication_date: '01-01-2020',
        multimedia: null,
        link: { url: '#' },
        byline: 'criticName',
        critics_pick: 0,
      },
    };
    const wrapper = mount(
      <Provider store={store}>
        <Review {...props} />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
