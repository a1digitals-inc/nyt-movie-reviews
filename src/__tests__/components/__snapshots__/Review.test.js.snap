// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Review component should render without crashing 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Review
    rawData={
      Object {
        "byline": "criticName",
        "critics_pick": 0,
        "display_title": "title",
        "link": Object {
          "url": "#",
        },
        "multimedia": null,
        "publication_date": "01-01-2020",
        "summary_short": "description",
      }
    }
  >
    <styled.div>
      <StyledComponent
        forwardedComponent={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-bdVaJa",
              "isStatic": false,
              "lastClassName": "bOtZPJ",
              "rules": Array [
                "
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding: 20px;
  margin-top: 15px;
  border-radius: 4px;
  background: #fff;

  &:hover {
    background: ",
                "#f5f5f5",
                ";
  }

  div {
    strong {
      display: block;
      font-size: 20px;
    }

    small {
      display: flex;
      align-items: stretch;
      padding: 0;
      margin: 2px 0 0 0;
      color: #999;

      img {
        width: 16px;
        height: 16px;
        padding: 0;
        margin: 0 4px 0 0;
      }
    }

    span {
      display: block;
      margin-top: 15px;
      color: #000000;
    }

    a {
      display: block;
      margin-top: 15px;
      color: #999;
      font-weight: bold;

      &:hover {
        color: ",
                "#191919",
                ";
      }
    }

    button {
      display: block;
      margin-top: 5px;
      color: #999;
      background-color: transparent;
      font-weight: bold;
      border: none;
      cursor: pointer;

      &:hover {
        color: ",
                "#191919",
                ";
      }
    }
  }

  img {
    margin: 0 0 0 20px;
    height: 120px;

    @media (max-width: 480px) {
      display: none;
    }
  }
",
              ],
            },
            "displayName": "styled.div",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-bdVaJa",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          }
        }
        forwardedRef={null}
      >
        <div
          className="sc-bdVaJa bOtZPJ"
        >
          <div>
            <strong>
              title
            </strong>
            <small>
              01/01/2020
            </small>
            <small>
              By 
              criticName
            </small>
            <span>
              description
            </span>
            <a
              href="#"
              rel="noopener noreferrer"
              target="_blank"
            >
              Read review
            </a>
            <button
              onClick={[Function]}
              type="button"
            >
              Add to favorites
            </button>
          </div>
          <img
            alt="MovieImage"
            src="empty-image.svg"
          />
        </div>
      </StyledComponent>
    </styled.div>
  </Review>
</Provider>
`;
